llvm/lib/Target/Primate/PrimateVLIWPacketizer.cpp

```diff
@@ -158,13 +220,49 @@ PrimatePacketizerList::addToPacket(MachineInstr &MI) {
 
 void PrimatePacketizerList::endPacket(MachineBasicBlock *MBB,
                                       MachineBasicBlock::iterator MI) {
+
+  // need to first generate the needed bypass ops
+  // generating bypass ops allows the fix up to x0 out the bypasses for free :>
+  MachineInstr* packet_breaking_instr = CurrentPacketMIs.back();
+  llvm::SmallVector<MachineInstr*, 6> generated_bypass_instrs;
+  MachineBasicBlock::iterator old_end_instr = MI;
+  bool push_branch_to_next_packet = false;
+  if(packet_breaking_instr->isBranch()) {
+    push_branch_to_next_packet = insertBypassOps(packet_breaking_instr, generated_bypass_instrs);
+  }
+
+  // if we push to the next packet then pop the branch from the back && set the isntr pointer back.
+  if(push_branch_to_next_packet) {
+    dbgs() << "pushing branch to a new packet.\n";
+    CurrentPacketMIs.pop_back();
+    --MI;
+    for(auto& _: generated_bypass_instrs){
+      --MI;
+      dbgs() << "reversing the MI iterator.\n";
+    }
+    MI->dump();
+  }
+  else if(packet_breaking_instr->isBranch() && generated_bypass_instrs.size() > 0) { // add bypass ops to the current packet MIs if we haven't pushed it off
+    CurrentPacketMIs.pop_back();
+    for(auto& bypass_op: generated_bypass_instrs) {
+      CurrentPacketMIs.push_back(bypass_op);
+    }
+    CurrentPacketMIs.push_back(packet_breaking_instr);
+    // really???
```

Comment:
This might be an instruction ordering issue.

Reply:
Part of fix in [5c29c54](https://github.com/FAST-Research-Group/primate-arch-gen/pull/1/commits/5c29c542e7c288e70fb7592a57b554905880ebe9)